drivers: display: Added Sharp memory display drivers.

rgujju opened the pull request:
Added support for sharp memory displays of the series LS0XX. (Previous PR was closed #27660)

This driver supports following displays
LS012B7DD01
LS012B7DD06
LS013B7DH03 -> Tested
LS013B7DH05
LS013B7DH06
LS027B7DH01A
LS032B7DD02
LS044Q7DH01

`extcomin-gpios` and `disp-gpios` are optional in device tree.

- If `disp-gpios` is provided then that GPIO will be set to high during initialization and can be controlled using the display blanking APIs.
- If `extcomin-gpios` is provided and `CONFIG_LS0XX_VCOM_DRIVER=y` EXTCOMIN pin will be used for toggling VCOM.
- If `CONFIG_LS0XX_VCOM_DRIVER=y` and extcomin-gpios is not provided then SPI (serial input) will be used for toggling VCOM.


**Note:** The CS control can not be done by SPI driver since it toggles CS after every line is written as seen in the image. This kind of CS control is not supported by the display.
![spi_transfer_with_CS](https://user-images.githubusercontent.com/10092220/100537873-11aa1b80-3252-11eb-91c4-31886c5d6cac.png)

rgujju committed:
drivers: display: Added support for restricted update.

Added support for restricted updating of monochrome
displays.

Signed-off-by: Rohit Gujarathi <gujju.rohit@gmail.com>

rgujju committed:
shields: display: Added LS013B7DH03 display shield.

Added sharp memory display LS013B7DH03 shield

Signed-off-by: Rohit Gujarathi <gujju.rohit@gmail.com>

rgujju committed:
drivers: display: Added Sharp memory display drivers.

Added support for sharp memory displays of the series
LS0XX.

Signed-off-by: Rohit Gujarathi <gujju.rohit@gmail.com>

vanwinkeljan commented:
@rgujju Where the open comments of #27660 addressed?

Further on the SPI CS why not just use the SPI driver with the flag `SPI_HOLD_ON_CS` set?
https://github.com/zephyrproject-rtos/zephyr/blob/6aa16e01dd919a7a434f402da95799be947dc6be/include/drivers/spi.h#L100-L104

rgujju commented:
@vanwinkeljan Yes they were!

> SPI_HOLD_ON_CS

Will check this out

rgujju commented:
@vanwinkeljan ,
Added CS control through SPI driver. Thanks for the info. 
I have also added ``extcomin-frequency`` in device tree so user can modify the EXTCOMIN pin toggling frequency. According to [app note](https://www.sharpsma.com/documents/1468207/1485747/Memory+LCD+Theory%2C+Programming%2C+and+Interfaces), 

> In general, for panels that require a 3 V power supply,the frequency of EXTCOMM is 60 Hz. For 5 V panels thefrequency is slower and more lenient, 1 – 20 Hz; again,check the Specifications. Higher frequency gives bettercontrast; lower frequency saves power.

jfischer-no commented on drivers/display/ls0xx.c:
```diff
@@ -0,0 +1,441 @@
+	driver->spi_config.operation |= SPI_TRANSFER_LSB;
+	driver->spi_config.operation |= SPI_CS_ACTIVE_HIGH | SPI_HOLD_ON_CS | SPI_LOCK_ON;
+
+	driver->cs_ctrl.gpio_dev = device_get_binding(
```
CS over gpio should be optional, else it will conflict with spi controller dedicated CS line, if any is available. see https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ethernet/eth_enc424j600.c#L616 as example.

rgujju replied:
@jfischer-no 
Which pin is to be used if user does not mention `cs-gpios`? Is it left to the user in application code to make CS line active?

jfischer-no replied:
Many SPI controllers have dedicated CS line, few not and uses GPIO for that, which is a bit slower. https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/spi.h#L124 , https://docs.zephyrproject.org/latest/reference/devicetree/api.html#spi

jfischer-no reviewed (changes requested):
Additional, checkpatch reports a lot of coding style issues. Fix it accordingly.

rgujju commented on drivers/display/Kconfig.ls0xx:
```diff
@@ -0,0 +1,20 @@
+	  VCOM will be toggled by driver based on LS0XX_EXTMODE_SERIAL, else
+	  user has to handle toggling. If LS0XX_VCOM_DRIVER=n and
+	  using serial mode then user has to handle contention between
+	  SPI transfers for write and VCOM toggle cmd.
```
@jfischer-no 
User would call the normal display write API.

rgujju commented on drivers/display/ls0xx.c:
```diff
@@ -0,0 +1,441 @@
+ */
+static int ls0xx_blanking_off(const struct device *dev)
+{
+#if DT_INST_NODE_HAS_PROP(0, disp_gpios)
```
@jfischer-no 
The manufacturer calls it the display pin and i have used the same naming here. What would you suggest?
This is optional because user can simply connect it to VDD to save a gpio pin.

rgujju replied:
what do you think about `disp_en_gpios`?

jfischer-no replied:
disp_en_gpios is probably better.

rgujju commented on drivers/display/ls0xx.c:
```diff
@@ -0,0 +1,441 @@
+
+#elif defined(CONFIG_LS0XX_VCOM_DRIVER) && !DT_INST_NODE_HAS_PROP(0, extcomin_gpios)
+static uint8_t extcomin_bit;
+#define LS0XX_EXTCOMIN_BIT extcomin_bit
```
@jfischer-no 
CONFIG_LS0XX_VCOM_DRIVER is required to enable VCOM toggling by driver. 
If user sets `CONFIG_LS0XX_VCOM_DRIVER=n` they have the freedom to use different methods for toggling to achieve lower power consumption for example schmitt trigger, 555, etc. Hence i suggest it should be kept.

jfischer-no replied:
Your code is unnecessarily obscuring. Please write it differently.

erwango commented:
> Additional, checkpatch reports a lot of coding style issues. Fix it accordingly.

@jfischer-no Did you run it manually? How happen this is not reported by CI bot anymore ?

jfischer-no commented:
> > Additional, checkpatch reports a lot of coding style issues. Fix it accordingly.
> 
> @jfischer-no Did you run it manually? How happen this is not reported by CI bot anymore ?

Yes I did. I am not sure what is wrong with CI, I thought we had an issue open about it, #29960

rgujju commented:
To simplify the code, i was thinking of removing the gpio toggling code completely from the driver and let user handle in application code. So this would eliminate `CONFIG_LS0XX_VCOM_DRIVER`, `extcomin-gpios` and `extcomin-frequency`. 

The driver would then toggle VCOM through SPI only.

What do you think @jfischer-no @vanwinkeljan ?

jfischer-no commented:
> To simplify the code, i was thinking of removing the gpio toggling code completely from the driver and let user handle in application code. This would allow the user more flexibility for toggling VCOM. So this would eliminate `CONFIG_LS0XX_VCOM_DRIVER`, `extcomin-gpios` and `extcomin-frequency`.
> 
> The driver would only toggle VCOM through SPI.

Easier would not be wrong, I have the 1.3 inch Waveshare display, give me a few days, I will try to find time to play with it.

vanwinkeljan commented on boards/shields/ls0xx_generic/Kconfig.defconfig:
```diff
@@ -0,0 +1,48 @@
+	string "Display device name"
+	default "LS0XX"
+
+config LVGL_HOR_RES
```
`LVGL_HOR_RES` has been renamed to `LVGL_HOR_RES_MAX`, update accordingly

vanwinkeljan commented on boards/shields/ls0xx_generic/Kconfig.defconfig:
```diff
@@ -0,0 +1,48 @@
+	int "Horizantal resolution"
+	default 128
+
+config LVGL_VER_RES
```
`LVGL_VER_RES` has been renamed to `LVGL_VER_RES_MAX`, update accordingly

vanwinkeljan commented on boards/shields/ls0xx_generic/doc/index.rst:
```diff
@@ -0,0 +1,122 @@
+.. _ls0xx_generic_shield:
+
+Sharp memory display generic shield
+#########################################
```
Length of this line should match the length of the line above.

vanwinkeljan commented on drivers/display/ls0xx.c:
```diff
@@ -0,0 +1,441 @@
+ *          Buffer width should be a multiple of 8
+ *
+ *          If CONFIG_LS0XX_VCOM_DRIVER=y and extcomin-gpios is not defined
+ *          and buffer is set to NULL, or height/width is 0 then
```
Comment is not correct  if width would be set to zero an error is returned, see line 240

vanwinkeljan commented on drivers/display/ls0xx.c:
```diff
@@ -0,0 +1,441 @@
+		return -EINVAL;
+	}
+
+	if ((x + desc->width) > LS0XX_PANEL_WIDTH) {
```
Is this actually not just checking if x is zero as line 240 enforces that width equals `LS0XX_PANEL_WIDTH `?

If this is the intent better just check if x == 0.

vanwinkeljan commented on include/drivers/display.h:
```diff
@@ -99,6 +103,15 @@ enum display_orientation {
  * Current display orientation
  *
+ * @var uint16_t display_capabilities::x_alignment
+ * Minimum update contraint in the x-direction-1.
```
typo: `contraint ` vs `constraint `

vanwinkeljan commented on include/drivers/display.h:
```diff
@@ -99,6 +103,15 @@ enum display_orientation {
+ * 0 means no contraint
+ *
+ * @var uint16_t display_capabilities::y_alignment
+ * Minimum update contraint in the y-direction-1.
```
typo: `contraint ` vs `constraint `

vanwinkeljan reviewed (changes requested):
Some minor remarks

And I'm fine with simplifying the code although I'm a bit concerned people will forget to toggle VCOM; but OK it is mentioned in the documentation.

> To simplify the code, i was thinking of removing the gpio toggling code completely from the driver and let user handle in application code. This would allow the user more flexibility for toggling VCOM. So this would eliminate CONFIG_LS0XX_VCOM_DRIVER, extcomin-gpios and extcomin-frequency.
>
> The driver would only toggle VCOM through SPI.

nixward commented:
> To simplify the code, i was thinking of removing the gpio toggling code completely from the driver and let user handle in application code. This would allow the user more flexibility for toggling VCOM. So this would eliminate `CONFIG_LS0XX_VCOM_DRIVER`, `extcomin-gpios` and `extcomin-frequency`.
> 

@rgujju 
It would be good to provide a default VCOM toggle that works out of the box but is optional.   The user can then could choose to turn this off and use another method from their application if they require.

Another low RAM alternative could be to use a PWM output at 50% duty cycle.

rgujju commented:
Hi @vanwinkeljan @jfischer-no 
I have updated the PR with requested changes, please have a look.
There is only one checkpatch warning for line length, i feel it improves readability. lmk if thats ok.

vanwinkeljan commented on boards/shields/ls0xx_generic/doc/index.rst:
```diff
@@ -0,0 +1,124 @@
+  called periodically for toggling VCOM.
+  If there is no data to update then buf can be set to NULL then
+  only VCOM will be toggled.
+  If using lvgl and there is no data to update `lv_task_handler`
```
Looks like CI doc build has problems with \`lv_task_handler\`, maybe use \* lv_task_handler\* instead.

rgujju replied:
done! @vanwinkeljan

jfischer-no commented on boards/shields/ls0xx_generic/doc/index.rst:
```diff
@@ -0,0 +1,124 @@
+  If there is no data to update then buf can be set to NULL then
+  only VCOM will be toggled.
+  If using lvgl and there is no data to update ``lv_task_handler``
+  will not call `display_write`. So explicit call to `display_write`
```
I do not think it is easy to achieve. The driver is not mutually exclusive in this regard, is it?

jfischer-no commented on boards/shields/ls0xx_generic/ls013b7dh03.overlay:
```diff
@@ -0,0 +1,22 @@
+		reg = <0>;
+		width = <128>;
+		height = <128>;
+		extcomin-gpios = <&gpio0 22 GPIO_ACTIVE_HIGH>; /* optional */
```
Must be arduino_header, for example `extcomin-gpios = <&arduino_header 14 GPIO_ACTIVE_HIGH>; /* D8 */`.
Remove /* optional */, that is already described in binding file.

jfischer-no commented on boards/shields/ls0xx_generic/ls013b7dh03.overlay:
```diff
@@ -0,0 +1,22 @@
+		height = <128>;
+		extcomin-gpios = <&gpio0 22 GPIO_ACTIVE_HIGH>; /* optional */
+		extcomin-frequency = <60>; /* required if extcomin-gpios is defined */
+		disp-en-gpios = <&gpio0 24 GPIO_ACTIVE_HIGH>; /* optional */
```
Must be arduino_header, for example  `disp-en-gpios = <&arduino_header 12 GPIO_ACTIVE_HIGH>;  /* D6 */`

rgujju commented on drivers/display/ls0xx.c:
```diff
@@ -0,0 +1,375 @@
+
+	LOG_DBG("Lines %d to %d", start_line, start_line + num_lines - 1);
+
+#if defined(CONFIG_LS0XX_VCOM_DRIVER) && \
```
@jfischer-no 
The only concern i have regarding this is what if the user requires the GPIO pin. Toggling via SPI would free up a GPIO for the user.

jfischer-no replied:
Well, I would recommend to choose a different display or use an I²C I/O Expander.

rgujju replied:
@jfischer-no 
Alright. Also, what do you think about using a pwm instead of gpio toggle?

nick-kraus replied:
I see the advantage of the _option_ to use a PWM pin, but I think requiring it isn't a great choice, as it will limit the applications in which you can use the display (in any setting where you are out of pwm pins, or just making pin selection a little less convenient).

Right now a user can choose to not define CONFIG_LS0XX_VCOM_DRIVER and set up a pwm pin to do that manually (if I understand correctly) without requiring all (extcomin based) applications to use a pwm resource.

jfischer-no replied:
> @jfischer-no
> Alright. Also, what do you think about using a pwm instead of gpio toggle?

@rgujju That could be added later, let's make it ready to be merged with minimal but clear functionality.

rgujju commented:
Hi @jfischer-no, @vanwinkeljan 
I have incorporated the suggested edits.

erwango reviewed (approved):
LGTM for shield part

vanwinkeljan commented on drivers/display/ls0xx.c:
```diff
@@ -0,0 +1,352 @@
+	tx_buf.buf =  &write_cmd;
+	err = spi_write(driver->spi_dev, &driver->spi_config, &buf_set);
+
+	if (buf != NULL) {
```
NULL check is no longer needed as SPI based VCOM is not supported anymore

vanwinkeljan commented on drivers/display/ls0xx.c:
```diff
@@ -0,0 +1,352 @@
+		tx_buf.buf = &line_buf[0];
+		for (int l = start_line; l <= start_line + num_lines - 1; l++) {
+			line_buf[0] = l;
+			memcpy(&line_buf[1], buf,
```
This memcpy and `line_buf` could be removed if two `spi_buf`s are used in `buf_set`

jfischer-no replied:
That is good point. I have worked it over [quickly](https://github.com/jfischer-no/zephyr/commit/75b2d31c1c95e4b93548897387a9222b4aa67544). You are free to take it over https://github.com/jfischer-no/zephyr/commit/75b2d31c1c95e4b93548897387a9222b4aa67544.patch and squash.

rgujju replied:
I had tried this earlier, but it affects the speed of the write. See below.
Separate buffers (suggested by you). It takes 18ms. The first transfer is the write cmd and then 3 buffers are transferred separately.
![image](https://user-images.githubusercontent.com/10092220/104937996-7a9d4c80-59d4-11eb-986f-c8f9de0bd9ed.png)

Single buffer. It takes 14ms.The first transfer is the write cmd and then the complete buffer is transferred.
![image](https://user-images.githubusercontent.com/10092220/104938149-ac161800-59d4-11eb-8026-6a9f821c54eb.png)

Both of these are tested on NRF52840 with DMA (SPIM peripheral) on LS013B7DH03. While 4ms may not seem high for a small display, it will quickly add up for a larger display and the lag would be perceivable.

jfischer-no replied:
I see difference of about 0.15ms for the update of 8 lines each 144px. The benefit is that there is no need for an intermediate buffer and the code is much simpler.

![Uphie7xe](https://user-images.githubusercontent.com/9864546/105103329-bc032a00-5ab0-11eb-8a19-931d0ade61ad.png)
![feeH2qua](https://user-images.githubusercontent.com/9864546/105103338-bf96b100-5ab0-11eb-88b7-83917bfb8cba.png)

rgujju replied:
The large difference is during updating the complete display. You can try with the example i used for toggling the complete display black and white. [(src)](https://github.com/rgujju/zephyr/blob/simple_display/samples/display/simple_display/src/main.c)

jfischer-no replied:
Well, regarding complete display update, depending on how much buffer (LVGL) is available it will scale according to the display size or be a little less. Decisive for this controller is the time it takes to write one line. The delay between LINE_NUMBER and DATA/DATA and DUMMY_BYTE will have less influence the larger the number of pixels in the line is.

Whatever you decide, with or without intermediate buffers, please rework the use of `spi_buf `and `spi_buf_set` and remove global `struct spi_buf tx_buf`and `struct spi_buf_set buf_set`.

rgujju commented on boards/shields/ls0xx_generic/ls013b7dh03.overlay:
```diff
@@ -0,0 +1,22 @@
+
+&arduino_spi {
+	status = "okay";
+	cs-gpios = <&arduino_header 11 GPIO_ACTIVE_HIGH>; /* D5 */
```
Can i still make changes and push?

rgujju replied:
Updated it!

jfischer-no replied:
Thanks
